// @stepzen/sdk/src/commands/deploy.ts
// Copyright IBM Corp. 2020, 2025

import debug from 'debug'
import fetch from '@stepzen/fetch'

import {ADMIN_DEPLOY_URL} from '../shared/constants.js'
import {getRequestHeaders} from '../shared/request.js'
import {
  SDKConfiguration,
  StepZenAccount,
  StepZenDeploy,
  ZenCtlResponse,
} from '../shared/types.js'

export default async (
  details: StepZenDeploy,
  account: StepZenAccount,
  sdkConfig: SDKConfiguration,
): Promise<ZenCtlResponse> => {
  const headers = getRequestHeaders(account, sdkConfig)

  const payload: any = {
    schema: `/workspaces/schemas/${details.schema}`,
  }

  if (details.configurationsets) {
    payload.configurationsets = details.configurationsets.map(
      config => `/workspaces/configurationsets/${config}`,
    )
  }

  debug('stepzen:headers')(headers)
  debug('stepzen:payload')(payload)

  const response = await fetch(
    `${account.server}${ADMIN_DEPLOY_URL}/${details.destination}`,
    {
      body: JSON.stringify(payload),
      headers: headers as any,
      method: 'POST',
    },
  )

  debug('stepzen:response')(response)

  if (response.status >= 400 && response.status < 500) {
    throw new Error(
      'Could not complete the request. Please check your authentication details are correct.',
    )
  }

  if (response.status >= 500) {
    return {
      errors: [response.statusText],
      success: false,
    }
  }

  const text = await response.text()
  let json

  try {
    json = JSON.parse(text)
  } catch {
    return {
      errors: [`An unexpected error occurred.\n\n${text}`],
      success: false,
    }
  }

  return json
}
